src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java

```diff
@@ -143,105 +131,94 @@ private SourcePaginations getPaginationStateForPlayer(Player source, boolean cre
-                .executor((src, args) -> {
-                    args.<ActivePagination>getOne("pagination-id").get().previousPage();
+        Command.Parameterized prev = Command.builder()
+                .setShortDescription(TextComponent.of("Go to the previous page"))
+                .setExecutor((context) -> {
+                    context.requireOne(paginationIdParameter).previousPage();
                     return CommandResult.success();
                 }).build();
 
-        CommandElement pageArgs = integer(t("page"));
+        Parameter.Value<Integer> pageParameter = integerNumber().setKey("page").build();
 
-        CommandExecutor pageExecutor = (src, args) -> {
-            args.<ActivePagination>getOne("pagination-id").get().specificPage(args.<Integer>getOne("page").get());
+        CommandExecutor pageExecutor = (context) -> {
+            context.requireOne(paginationIdParameter).specificPage(context.requireOne(pageParameter));
             return CommandResult.success();
         };
 
-        CommandSpec page = CommandSpec.builder()
-                        .description(t("Go to a specific page"))
-                        .arguments(pageArgs)
-                        .executor(pageExecutor).build();
-
-        //Fallback to page arguments
-        ChildCommandElementExecutor childDispatcher = new ChildCommandElementExecutor(pageExecutor);
-        childDispatcher.register(next, "next", "n");
-        childDispatcher.register(prev, "prev", "p", "previous");
-        childDispatcher.register(page, "page");
+        Command.Parameterized page = Command.builder()
+                .setShortDescription(TextComponent.of("Go to a specific page"))
+                .parameter(pageParameter)
+                .setExecutor(pageExecutor)
+                .build();
 
         //We create the child manually in order to force that paginationElement is required for all children + fallback
         //https://github.com/SpongePowered/SpongeAPI/issues/1272
-        return CommandSpec.builder().arguments(paginationElement, firstParsing(childDispatcher, pageArgs))
-                .executor(childDispatcher)
-                .description(t("Helper command for paginations occurring"))
+        return Command.builder()
+                .parameters(paginationIdParameter, firstOf(pageParameter,
+                        subcommand(next, "next", "n"),
+                        subcommand(prev, "prev", "p", "previous"),
+                        subcommand(page, "page")))
+                .child(page, "page")
+                .setExecutor(page)
+                .setShortDescription(TextComponent.of("Helper command for paginations occurring"))
                 .build();
     }
 
-    private class ActivePaginationCommandElement extends CommandElement {
-
-        protected ActivePaginationCommandElement(@Nullable Text key) {
-            super(key);
-        }
+    private class ActivePaginationParser implements ValueParser<ActivePagination> {
```

Comment:
Any reason that this is split into `ValueParser` and `ValueCompleter`? In general, they should just be one `ValueParameter` if this is the case ([the docs say if a `ValueParameter` is passed it will use the completer there](https://github.com/SpongePowered/SpongeAPI/blob/api-8/src/main/java/org/spongepowered/api/command/parameter/Parameter.java#L979-L983)).

Reply:
I missed that part, will do !